File: solutions/0557. Reverse Words in a String III/0557.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
public:
    string reverseWords(string &s) {
        string::iterator wordStart = s.begin();
        for (string::iterator wordEnd = s.begin(); wordEnd != s.end(); ++wordEnd) {
            if (*wordEnd == ' ') {
                reverse(wordStart, wordEnd);
                wordStart = wordEnd;
                ++wordStart;
            }
        }
        reverse(wordStart, s.end());
        return move(s);
    }
};

File: solutions/0557. Reverse Words in a String III/0557.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
  def reverseWords(self, s: str) -> str:
    s = list(s)
    i = 0
    j = 0
    n = len(s)

    while i < n:
      while i < j or (i < n and s[i] == ' '):
        i += 1
      while j < i or (j < n and s[j] != ' '):
        j += 1
      s[i:j] = reversed(s[i:j])

    return ''.join(s)